File: apps/api/app/schemas/blog.py
"""
Blog post schemas for request/response validation.
"""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class BlogPostBase(BaseModel):
    """Base schema for blog posts."""

    title: str = Field(..., min_length=1, max_length=255)
    subtitle: str | None = Field(None, max_length=500)
    content: str = Field(..., min_length=1)
    excerpt: str = Field(..., min_length=1, max_length=300)
    category: str = Field(..., min_length=1, max_length=50)
    tags: list[str] = Field(default_factory=list)
    featured_image_url: str | None = Field(None, max_length=500)
    seo_title: str | None = Field(None, max_length=70)
    seo_description: str | None = Field(None, max_length=160)
    seo_keywords: list[str] | None = None


class BlogPostCreate(BlogPostBase):
    """Schema for creating a blog post."""

    slug: str | None = Field(None, max_length=255)
    published: bool = False


class BlogPostUpdate(BaseModel):
    """Schema for updating a blog post."""

    title: str | None = Field(None, min_length=1, max_length=255)
    subtitle: str | None = Field(None, max_length=500)
    content: str | None = Field(None, min_length=1)
    excerpt: str | None = Field(None, min_length=1, max_length=300)
    category: str | None = Field(None, min_length=1, max_length=50)
    tags: list[str] | None = None
    featured_image_url: str | None = Field(None, max_length=500)
    seo_title: str | None = Field(None, max_length=70)
    seo_description: str | None = Field(None, max_length=160)
    seo_keywords: list[str] | None = None
    published: bool | None = None


class AuthorInfo(BaseModel):
    """Author information for blog posts."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    full_name: str | None = None


class BlogPostRead(BaseModel):
    """Schema for reading a blog post (public view)."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    slug: str
    title: str
    subtitle: str | None
    content: str
    excerpt: str
    category: str
    tags: list[str]
    featured_image_url: str | None
    seo_title: str | None
    seo_description: str | None
    seo_keywords: list[str] | None
    read_time_minutes: int
    views_count: int
    published_at: datetime | None
    created_at: datetime
    updated_at: datetime
    author: AuthorInfo | None = None

    @property
    def effective_seo_title(self) -> str:
        """Get the SEO title or fallback to regular title."""
        return self.seo_title or self.title

    @property
    def effective_seo_description(self) -> str:
        """Get the SEO description or fallback to excerpt."""
        return self.seo_description or self.excerpt


class BlogPostListItem(BaseModel):
    """Schema for blog post list item (summary view)."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    slug: str
    title: str
    subtitle: str | None
    excerpt: str
    category: str
    tags: list[str]
    featured_image_url: str | None
    read_time_minutes: int
    views_count: int
    published_at: datetime | None
    created_at: datetime
    author: AuthorInfo | None = None


class BlogPostListResponse(BaseModel):
    """Response schema for paginated blog post list."""

    items: list[BlogPostListItem]
    total: int
    page: int
    page_size: int
    total_pages: int


class BlogCategoryCount(BaseModel):
    """Category with post count."""

    category: str
    count: int


class BlogTagCount(BaseModel):
    """Tag with post count."""

    tag: str
    count: int


class BlogMetadata(BaseModel):
    """Blog metadata for sidebar/navigation."""

    categories: list[BlogCategoryCount]
    popular_tags: list[BlogTagCount]
    total_posts: int

File: apps/web/src/services/blog.ts
/**
 * Blog service for API communication.
 */

import { apiClient } from './api';

export interface AuthorInfo {
  id: string;
  full_name: string | null;
}

export interface TranslationInfo {
  locale: string;
  slug: string;
  title: string;
}

export interface BlogPost {
  id: string;
  slug: string;
  locale: string;
  translation_key: string | null;
  title: string;
  subtitle: string | null;
  content: string;
  excerpt: string;
  category: string;
  tags: string[];
  featured_image_url: string | null;
  seo_title: string | null;
  seo_description: string | null;
  seo_keywords: string[] | null;
  read_time_minutes: number;
  views_count: number;
  is_featured: boolean;
  published_at: string | null;
  created_at: string;
  updated_at: string;
  author: AuthorInfo | null;
  available_translations: TranslationInfo[] | null;
}

export interface BlogPostListItem {
  id: string;
  slug: string;
  locale: string;
  translation_key: string | null;
  title: string;
  subtitle: string | null;
  excerpt: string;
  category: string;
  tags: string[];
  featured_image_url: string | null;
  read_time_minutes: number;
  views_count: number;
  is_featured: boolean;
  published_at: string | null;
  created_at: string;
  author: AuthorInfo | null;
}

export interface BlogPostListResponse {
  items: BlogPostListItem[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface BlogCategoryCount {
  category: string;
  count: number;
}

export interface BlogTagCount {
  tag: string;
  count: number;
}

export interface BlogMetadata {
  categories: BlogCategoryCount[];
  popular_tags: BlogTagCount[];
  total_posts: number;
}

export interface GetPostsParams {
  page?: number;
  page_size?: number;
  category?: string;
  tag?: string;
  locale?: string;
}

/**
 * Get paginated list of published blog posts.
 */
export async function getBlogPosts(params: GetPostsParams = {}): Promise<BlogPostListResponse> {
  const searchParams = new URLSearchParams();

  if (params.page) searchParams.set('page', params.page.toString());
  if (params.page_size) searchParams.set('page_size', params.page_size.toString());
  if (params.category) searchParams.set('category', params.category);
  if (params.tag) searchParams.set('tag', params.tag);
  if (params.locale) searchParams.set('locale', params.locale);

  const queryString = searchParams.toString();
  const url = queryString ? `/api/v1/blog/posts?${queryString}` : '/api/v1/blog/posts';

  return apiClient.get<BlogPostListResponse>(url);
}

/**
 * Get a single blog post by slug.
 */
export async function getBlogPost(slug: string, locale?: string): Promise<BlogPost> {
  const searchParams = new URLSearchParams();
  if (locale) searchParams.set('locale', locale);
  const queryString = searchParams.toString();
  const url = queryString
    ? `/api/v1/blog/posts/${slug}?${queryString}`
    : `/api/v1/blog/posts/${slug}`;
  return apiClient.get<BlogPost>(url);
}

/**
 * Get blog metadata (categories, tags, total posts).
 */
export async function getBlogMetadata(locale?: string): Promise<BlogMetadata> {
  const searchParams = new URLSearchParams();
  if (locale) searchParams.set('locale', locale);
  const queryString = searchParams.toString();
  const url = queryString ? `/api/v1/blog/metadata?${queryString}` : '/api/v1/blog/metadata';
  return apiClient.get<BlogMetadata>(url);
}

/**
 * Get recent blog posts.
 */
export async function getRecentPosts(
  limit: number = 5,
  locale?: string
): Promise<BlogPostListItem[]> {
  const searchParams = new URLSearchParams();
  searchParams.set('limit', limit.toString());
  if (locale) searchParams.set('locale', locale);
  return apiClient.get<BlogPostListItem[]>(`/api/v1/blog/recent?${searchParams.toString()}`);
}
